lasagne/layers/recurrent.py

```diff
@@ -58,70 +97,44 @@ def __init__(self, input_layer, input_to_hidden, hidden_to_hidden,
 
         # Get the batch size and number of units based on the expected output
         # of the input-to-hidden layer
-        (n_batch, self.num_units) = self.input_to_hidden.get_output_shape()
+        (self.n_batch, _, self.num_inputs) = self.input_shape
+        self.num_units = i2h_out[-1]
 
         # Initialize hidden state
-        self.hid_init = self.create_param(hid_init, (n_batch, self.num_units))
+        self.hid_init = self.add_param(hid_init, (1, self.num_units),
+                                       trainable=learn_init, name="hid_init")
```

Comment:
Note that this parameter is now tagged as `regularizable`, because that's the default. Is this intended? I.e., when you train an RNN with L2 weight decay and with `learn_init=True`, do you want to constrain the norm of the initialization vector or not? I'm not that familiar with the RNN literature, but my gut feeling would be that you don't regularize it. The initialization vector should provide a neutral hidden state to start from, and the hidden state of an RNN includes an unregularized bias vector, so at least for ReLUs, it could be arbitrarily large.
You could also set `regularizable=False` and `init_state=True`, so people can choose whether or not to regularize them by providing appropriate tags in the `get_all_params()` calls.